packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -158,12 +158,25 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
             return;
         }
 
-        this.model.applyEdits([this.p2m.asTextEdit({
+        this.applyEdits([this.p2m.asTextEdit({
             range: this.m2p.asRange(this.model.getFullModelRange()),
             newText
         }) as monaco.editor.IIdentifiedSingleEditOperation]);
     }
+
+    protected ignoreEdits = false;
+    protected applyEdits(operations: monaco.editor.IIdentifiedSingleEditOperation[]): monaco.editor.IIdentifiedSingleEditOperation[] {
```

Comment:
I don't understand that one , is it that applyEdits is async somehow?

Reply:
Yes, it is performed either by a fs change or during a save operation and consequently should not make document dirty and trigger a new save. (the document is already saved or will be saved).

Reply:
Ha yes... I wonder if it would be easier to queue the save & sync methods then adding 3 mutexs  :(

Reply:
I like an idea of having a queue of cancellable sync and save operations, then there should be no chance to have a race between them. I will give it a try.

What do you mean instead of 3 mutexs? I cannot replace 2 cancellation tokens by one since a sync operation can be canceled by fs changes as well as by document changes and a save operation can be canceled only by document changes.

Reply:
Yes I did not mean replace it with one... I just meant the current structure with 2 cancellation tokens and one mutex is hard to follow. 

Seems like we need to serialize this better like with the queue idea this way we can treat theses operations as transactions ie. only remove them from the queue once all async/await calls are done for that operation and ensure that only one operation executes at a time.

Seems like it could be a useful abstraction in general too :)